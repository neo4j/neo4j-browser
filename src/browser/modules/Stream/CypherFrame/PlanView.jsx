/*
 * Copyright (c) 2002-2019 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { Component } from 'react'
import { PlanSVG } from './PlanView.styled'
import { dim } from 'browser-styles/constants'
import { deepEquals, shallowEquals } from 'services/utils'
import bolt from 'services/bolt/bolt'
import { FrameButton } from 'browser-components/buttons'
import { DoubleUpIcon, DoubleDownIcon } from 'browser-components/icons/Icons'
import {
  StyledOneRowStatsBar,
  StyledRightPartial,
  StyledLeftPartial
} from '../styled'
import { StyledFrameTitlebarButtonSection } from 'browser/modules/Frame/styled'
import Ellipsis from 'browser-components/Ellipsis'
import queryPlan from '../../D3Visualization/lib/visualization/components/queryPlan'

export class PlanView extends Component {
  constructor(props) {
    super(props)
    this.state = {
      extractedPlan: null
    }
  }

  componentDidMount() {
    this.extractPlan(this.props.result).catch(e => {})
  }

  componentWillReceiveProps(props) {
    if (props.updated !== this.props.updated) {
      return this.extractPlan(props.result || {})
        .then(() => {
          this.ensureToggleExpand(props)
        })
        .catch(e => {
          console.log(e)
        })
    }
    this.ensureToggleExpand(props)
    props.assignVisElement && props.assignVisElement(this.el, this.plan)
  }

  shouldComponentUpdate(props, state) {
    if (this.props.result === undefined) return true
    return (
      !deepEquals(props.result.summary, this.props.result.summary) ||
      !shallowEquals(state, this.state) ||
      props._planExpand !== this.props._planExpand
    )
  }

  extractPlan(result) {
    if (result === undefined) return Promise.reject(new Error('No result'))
    return new Promise((resolve, reject) => {
      const extractedPlan = bolt.extractPlan(result)
      if (extractedPlan) return this.setState({ extractedPlan }, resolve())
      resolve()
    })
  }

  planInit(el) {
    if (el != null && !this.plan) {
      const NeoConstructor = queryPlan
      this.el = el
      this.plan = new NeoConstructor(this.el)
      this.plan.display(this.state.extractedPlan)
      this.plan.boundingBox = () => {
        return this.el.getBBox()
      }

      this.props.assignVisElement &&
        this.props.assignVisElement(this.el, this.plan)
    }
  }

  ensureToggleExpand(props) {
    if (props._planExpand && props._planExpand !== this.props._planExpand) {
      switch (props._planExpand) {
        case 'COLLAPSE': {
          this.toggleExpanded(false)
          break
        }
        case 'EXPAND': {
          this.toggleExpanded(true)
          break
        }
      }
    }
  }

  toggleExpanded(expanded) {
    const visit = operator => {
      operator.expanded = expanded
      if (operator.children) {
        operator.children.forEach(child => {
          visit(child)
        })
      }
    }
    const tmpPlan = { ...this.state.extractedPlan }
    visit(tmpPlan.root)
    this.plan.display(tmpPlan)
  }

  render() {
    if (!this.state.extractedPlan) return null
    return (
      <PlanSVG
        data-testid="planSvg"
        style={
          this.props.fullscreen
            ? { 'padding-bottom': dim.frameStatusbarHeight + 'px' }
            : {}
        }
        ref={this.planInit.bind(this)}
      />
    )
  }
}

export class PlanStatusbar extends Component {
  state = {
    extractedPlan: null
  }

  componentDidMount() {
    if (this.props === undefined || this.props.result === undefined) return
    const extractedPlan = bolt.extractPlan(this.props.result, true)
    if (extractedPlan) this.setState({ extractedPlan })
  }

  componentWillReceiveProps(props) {
    if (props.result === undefined) return
    if (
      this.props.result === undefined ||
      !deepEquals(props.result.summary, this.props.result.summary)
    ) {
      const extractedPlan = bolt.extractPlan(props.result, true)
      this.setState({ extractedPlan })
    }
  }

  shouldComponentUpdate(props, state) {
    if (this.props.result === undefined) return true
    return !deepEquals(state, this.state)
  }

  render() {
    const plan = this.state.extractedPlan
    if (!plan) return null
    const { result = {} } = this.props
    return (
      <StyledOneRowStatsBar>
        <StyledLeftPartial>
          <Ellipsis>
            Cypher version: {plan.root.version}, planner: {plan.root.planner},
            runtime: {plan.root.runtime}.
            {plan.root.totalDbHits
              ? ` ${
                  plan.root.totalDbHits
                } total db hits in ${result.summary.resultAvailableAfter
                  .add(result.summary.resultConsumedAfter)
                  .toNumber() || 0} ms.`
              : ''}
          </Ellipsis>
        </StyledLeftPartial>
        <StyledRightPartial>
          <StyledFrameTitlebarButtonSection>
            <FrameButton
              data-testid="planCollapseButton"
              onClick={() =>
                this.props.setParentState({ _planExpand: 'COLLAPSE' })
              }
            >
              <DoubleUpIcon />
            </FrameButton>
            <FrameButton
              data-testid="planExpandButton"
              onClick={() =>
                this.props.setParentState({ _planExpand: 'EXPAND' })
              }
            >
              <DoubleDownIcon />
            </FrameButton>
          </StyledFrameTitlebarButtonSection>
        </StyledRightPartial>
      </StyledOneRowStatsBar>
    )
  }
}
