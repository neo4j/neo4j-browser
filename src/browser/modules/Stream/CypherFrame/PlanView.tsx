/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import memoize from 'memoize-one'
import React, { Component } from 'react'

import queryPlan from '../../D3Visualization/lib/visualization/components/queryPlan'
import {
  StyledLeftPartial,
  StyledOneRowStatsBar,
  StyledRightPartial
} from '../styled'
import { PlanExpand } from './CypherFrame'
import { PlanSVG } from './PlanView.styled'
import Ellipsis from 'browser-components/Ellipsis'
import { FrameButton } from 'browser-components/buttons'
import { DoubleDownIcon, DoubleUpIcon } from 'browser-components/icons/Icons'
import { dim } from 'browser-styles/constants'
import { StyledFrameTitlebarButtonSection } from 'browser/modules/Frame/styled'
import bolt from 'services/bolt/bolt'
import { deepEquals, shallowEquals } from 'services/utils'

type PlanViewState = { extractedPlan: any }
export type PlanViewProps = {
  planExpand: PlanExpand
  setPlanExpand: (p: PlanExpand) => void
  result: any
  updated: any
  assignVisElement: (a: any, b: any) => void
  isFullscreen: boolean
}

export class PlanView extends Component<PlanViewProps, PlanViewState> {
  el: any
  plan: any
  constructor(props: any) {
    super(props)
    this.state = {
      extractedPlan: null
    }
  }

  componentDidMount() {
    this.extractPlan(this.props.result)
      .then(() => {
        this.props.setPlanExpand('EXPAND')
        this.toggleExpanded(true)
      })
      .catch(() => {})
  }

  componentDidUpdate(prevProps: any): any {
    if (prevProps.updated !== this.props.updated) {
      return this.extractPlan(this.props.result || {})
        .then(() => {
          this.ensureToggleExpand(prevProps)
        })
        .catch(e => {
          console.log(e)
        })
    }
    this.ensureToggleExpand(prevProps)
    this.props.assignVisElement &&
      this.props.assignVisElement(this.el, this.plan)
  }

  shouldComponentUpdate(props: PlanViewProps, state: PlanViewState) {
    if (this.props.result === undefined) return true
    return (
      props.isFullscreen !== this.props.isFullscreen ||
      !deepEquals(props.result.summary, this.props.result.summary) ||
      !shallowEquals(state, this.state) ||
      props.planExpand !== this.props.planExpand
    )
  }

  extractPlan(result: any) {
    if (result === undefined) return Promise.reject(new Error('No result'))
    return new Promise<void>(resolve => {
      const extractedPlan = bolt.extractPlan(result)
      if (extractedPlan)
        return this.setState({ extractedPlan }, resolve() as any)
      resolve()
    })
  }

  planInit(el: any) {
    if (el != null && !this.plan) {
      const NeoConstructor: any = queryPlan
      this.el = el
      this.plan = new NeoConstructor(this.el)
      this.plan.display(this.state.extractedPlan)
      this.plan.boundingBox = () => {
        return this.el.getBBox()
      }

      this.props.assignVisElement &&
        this.props.assignVisElement(this.el, this.plan)
    }
  }

  ensureToggleExpand(prevProps: any) {
    if (
      this.props.planExpand &&
      this.props.planExpand !== prevProps.planExpand
    ) {
      switch (this.props.planExpand) {
        case 'COLLAPSE': {
          this.toggleExpanded(false)
          break
        }
        case 'EXPAND': {
          this.toggleExpanded(true)
          break
        }
      }
    }
  }

  toggleExpanded(expanded: any) {
    const visit = (operator: any) => {
      operator.expanded = expanded
      if (operator.children) {
        operator.children.forEach((child: any) => {
          visit(child)
        })
      }
    }
    const tmpPlan = { ...this.state.extractedPlan }
    visit(tmpPlan.root)
    this.plan.display(tmpPlan)
  }

  render() {
    if (!this.state.extractedPlan) return null
    return (
      <PlanSVG
        data-testid="planSvg"
        style={
          this.props.isFullscreen
            ? // @ts-expect-error ts-migrate(2769) FIXME: Object literal may only specify known properties, ... Remove this comment to see the full error message
              { 'padding-bottom': dim.frameStatusbarHeight + 'px' }
            : {}
        }
        ref={this.planInit.bind(this)}
      />
    )
  }
}

type PlanStatusbarProps = {
  result: any
  setPlanExpand: (p: PlanExpand) => void
}

const extractMemoizedPlan = memoize(
  result => bolt.extractPlan(result, true),
  (newArgs: any[], lastArgs: any[]) =>
    deepEquals(newArgs[0]?.summary, lastArgs[0]?.summary)
)

export function PlanStatusbar(props: PlanStatusbarProps) {
  const { result } = props
  if (!result || !result.summary) return null

  const plan = extractMemoizedPlan(result)
  if (!plan) return null

  return (
    <StyledOneRowStatsBar>
      <StyledLeftPartial>
        <Ellipsis>
          Cypher version: {plan.root.version}, planner: {plan.root.planner},
          runtime: {plan.root.runtime}.
          {plan.root.totalDbHits
            ? ` ${plan.root.totalDbHits} total db hits in ${
                result.summary.resultAvailableAfter
                  .add(result.summary.resultConsumedAfter)
                  .toNumber() || 0
              } ms.`
            : ''}
        </Ellipsis>
      </StyledLeftPartial>
      <StyledRightPartial>
        <StyledFrameTitlebarButtonSection>
          <FrameButton
            title="Collapse Plan"
            dataTestId="planCollapseButton"
            onClick={() => props.setPlanExpand('COLLAPSE')}
          >
            <DoubleUpIcon />
          </FrameButton>
          <FrameButton
            dataTestId="planExpandButton"
            title="Expand Plan"
            onClick={() => props.setPlanExpand('EXPAND')}
          >
            <DoubleDownIcon />
          </FrameButton>
        </StyledFrameTitlebarButtonSection>
      </StyledRightPartial>
    </StyledOneRowStatsBar>
  )
}
