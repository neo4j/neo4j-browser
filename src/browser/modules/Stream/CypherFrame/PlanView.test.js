/*
 * Copyright (c) 2002-2019 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/* global describe, test, expect */
import React from 'react'
import { render } from '@testing-library/react/pure'
import { v1 as neo4j } from 'neo4j-driver'
import { PlanView, PlanStatusbar } from './PlanView'

describe('PlanViews', () => {
  describe('PlanView', () => {
    test('displays plan view if it exists', () => {
      // Given
      const props = {
        query: 'MATCH xx0',
        result: {
          summary: {
            plan: {
              dbHits: 'xx0',
              arguments: {},
              children: [],
              operatorType: 'ProduceResults',
              identifiers: ['n']
            }
          }
        }
      }

      // When
      const { getByText } = render(<PlanView {...props} />)

      // Then
      expect(getByText('ProduceResults'))
    })
  })
  describe('PlanStatusbar', () => {
    test('displays statusBarMessage', () => {
      // Given
      const props = {
        result: {
          summary: {
            resultAvailableAfter: neo4j.int(100),
            resultConsumedAfter: neo4j.int(20),
            profile: {
              children: [],
              arguments: {
                operatorType: {},
                version: 'xx0',
                planner: 'xx1',
                runtime: 'xx2',
                children: []
              },
              dbHits: 'xx3'
            }
          }
        }
      }

      // When
      const { getByText } = render(<PlanStatusbar {...props} />)

      // Then
      expect(getByText(/xx0/))
      expect(getByText(/xx1/))
      expect(getByText(/xx2/))
      expect(getByText(/xx3/))
    })
  })
})
