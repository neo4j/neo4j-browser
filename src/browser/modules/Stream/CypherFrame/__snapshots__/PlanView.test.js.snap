// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanViews PlanStatusbar displays statusBarMessage 1`] = `
<div>
  <div
    class="styled__StyledStatsBar-sc-1dtvgs1-32 styled__StyledOneRowStatsBar-sc-1dtvgs1-33 dkmnzU"
  >
    <div
      class="styled__StyledLeftPartial-sc-1dtvgs1-42 cEQgJw"
    >
      <div
        class="Ellipsis-sc-1dn9bou-0 gzqRPC"
      >
        Cypher version: 
        xx0
        , planner: 
        xx1
        , runtime: 
        xx2
        .
         xx3 total db hits in 120 ms.
      </div>
    </div>
    <div
      class="styled__StyledRightPartial-sc-1dtvgs1-41 zWSxr"
    >
      <ul
        class="styled__FrameTitlebarButtonSection-sc-1dtvgs1-10 dRnwVT"
      >
        <li
          class="buttons__StyledFrameButton-sc-3xi0d5-10 cajKMw"
          data-testid="planCollapseButton"
        >
          <i
            class="sl-double-up IconContainer__I-sc-16whwc5-0 bXwiKG"
          />
        </li>
        <li
          class="buttons__StyledFrameButton-sc-3xi0d5-10 cajKMw"
          data-testid="planExpandButton"
        >
          <i
            class="sl-double-down IconContainer__I-sc-16whwc5-0 bXwiKG"
          />
        </li>
      </ul>
    </div>
  </div>
</div>
`;
